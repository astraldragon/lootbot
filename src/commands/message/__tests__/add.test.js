jest.mock("discord-akairo");
jest.mock("../../../models", () => {
  return {
    Message: { findOrCreate: jest.fn("findOrCreate") },
    Tier: { findOne: jest.fn("findOne") }
  };
});

const { Command } = require("discord-akairo");
const { Message, Tier } = require("../../../models");
const { checkManagePermissions } = require("../../../support");
const AddCommand = require("../add");

describe("commands/message/add", () => {
  beforeEach(() => {
    this.command = new AddCommand();

    this.args = {
      name: "msg1",
      message: "this is a message",
      type: "intro",
      delay: null,
      tier: null,
      user: null
    };

    this.tier = { id: 1, name: "Common" };
    this.user = { id: 1 };

    this.msg = {
      guild: { id: "1234" },
      channel: {
        send: jest.fn("send").mockReturnValue(Promise.resolve())
      }
    };
  });

  it("should configure properly", () => {
    expect(Command).toMatchSnapshot();
  });

  it("should validate the type field", () => {
    expect(Command.mock.calls[0][1].args[2].validate("nonvalid")).toBe(false);
    expect(Command.mock.calls[0][1].args[2].validate("intro")).toBe(true);
    expect(Command.mock.calls[0][1].args[2].validate("draw")).toBe(true);
    expect(Command.mock.calls[0][1].args[2].validate("reward")).toBe(true);
  });

  it("should have the proper permission check", () => {
    expect(Command.mock.calls[0][1].options.permissions).toBe(
      checkManagePermissions
    );
  });

  it("should add only non-null values", async () => {
    Message.findOrCreate.mockReturnValue([{ id: 1 }, true]);
    this.args.delay = 10;
    await this.command.exec(this.msg, this.args);

    let defaults = { ...this.args, guild: this.msg.guild.id };
    delete defaults.tier;
    delete defaults.user;

    expect(Message.findOrCreate).toHaveBeenCalledWith({
      where: { name: this.args.name, guild: this.msg.guild.id },
      defaults
    });

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} added`
    );
  });

  it("should query for the passed in tier", () => {
    this.args.tier = this.tier.name;
    this.command.exec(this.msg, this.args);

    expect(Tier.findOne).toHaveBeenCalledWith({
      where: { name: this.tier.name, guild: this.msg.guild.id }
    });
  });

  it("should notify when a tier isn't found", async () => {
    Tier.findOne.mockReturnValue(null);
    this.args.tier = this.tier.name;
    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      "A valid tier must be provided."
    );
  });

  it("should add the message with found tier", async () => {
    Tier.findOne.mockReturnValue(this.tier);
    Message.findOrCreate.mockReturnValue([{ id: 1 }, true]);

    this.args.tier = this.tier.name;
    await this.command.exec(this.msg, this.args);

    let defaults = {
      ...this.args,
      guild: this.msg.guild.id,
      tier_id: this.tier.id
    };

    delete defaults.tier;
    delete defaults.delay;
    delete defaults.user;

    expect(Message.findOrCreate).toHaveBeenCalledWith({
      where: { name: this.args.name, guild: this.msg.guild.id },
      defaults
    });

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} added`
    );
  });

  it("should add a user if provided", async () => {
    Tier.findOne.mockReturnValue(this.tier);
    Message.findOrCreate.mockReturnValue([{ id: 1 }, true]);

    this.args.user = this.user;
    await this.command.exec(this.msg, this.args);

    let defaults = {
      ...this.args,
      guild: this.msg.guild.id,
      user_id: this.user.id
    };

    delete defaults.tier;
    delete defaults.delay;
    delete defaults.user;

    expect(Message.findOrCreate).toHaveBeenCalledWith({
      where: { name: this.args.name, guild: this.msg.guild.id },
      defaults
    });

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} added`
    );
  });

  it("should notify when the message added already exists", async () => {
    Message.findOrCreate.mockReturnValue([{ id: 1 }, false]);
    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} already exists`
    );
  });

  it("should notify if any errors occur during the command", async () => {
    Message.findOrCreate.mockImplementation(() => {
      throw new Error("test");
    });

    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `An error occurred adding ${this.args.name}`
    );
  });
});
