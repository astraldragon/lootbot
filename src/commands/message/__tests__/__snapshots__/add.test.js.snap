// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/message/add should configure properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "message-add",
      Object {
        "aliases": Array [
          "message-add",
          "ma",
        ],
        "args": Array [
          Object {
            "id": "name",
            "prompt": Object {
              "start": "What is the identier of the message?",
            },
            "type": "string",
          },
          Object {
            "id": "message",
            "prompt": Object {
              "start": "What is the message?",
            },
            "type": "string",
          },
          Object {
            "id": "type",
            "prompt": Object {
              "start": "What is the type of the message? (intro, draw, or reward)",
            },
            "type": "string",
            "validate": [Function],
          },
          Object {
            "default": null,
            "id": "delay",
            "match": "prefix",
            "prefix": "delay=",
            "prompt": Object {
              "optional": true,
              "start": "What is the delay until the next message is shown?",
            },
            "type": "integer",
          },
          Object {
            "default": null,
            "id": "tier",
            "match": "prefix",
            "prefix": "tier=",
            "prompt": Object {
              "optional": true,
              "start": "What is the tier of the loot? (blank will be included in all tiers of loot)",
            },
            "type": "string",
          },
          Object {
            "default": null,
            "id": "user",
            "match": "prefix",
            "prefix": "user=",
            "prompt": Object {
              "optional": true,
              "start": "What user should receive this message? (blank will be all users)",
            },
            "type": "member",
          },
        ],
        "category": "Message",
        "channelRestriction": "guild",
        "description": Object {
          "content": "Add message to the lootbox. Messages can be targeted at specific users. Messages can also be targeted at specific tiers of loot (draw and reward types only)",
          "examples": Array [
            "message-add basicintro \\"Some intro...\\" intro",
            "message-add basicdraw \\"You won a prize!\\" draw",
            "message-add drawtier \\"<tier> prize...womp womp\\" draw tier=Common",
            "message-add drawtierdelay \\"<tier> prize...womp womp. Waiting...\\" draw delay=10 tier=Common",
            "message-add drawtieruser \\"Specific user, you have won a <tier> prize...womp womp\\" draw tier=Common user=@user",
            "message-add basicreward \\"Congrats <user>, you won loot named <reward>\\" reward",
            "message-add rewardtier \\"Congrats on your amazing prize\\" reward tier=Legendary",
            "message-add rewardtieruser \\"Specific user, congrats\\" reward user=@user",
          ],
          "usage": "<name> <message> <type> delay=<delay> tier=<tier> user=<user>",
        },
        "options": Object {
          "permissions": [Function],
        },
        "split": "sticky",
      },
    ],
  ],
}
`;
